### PR TITLE
Add config for dumping json and preprocessed ir

### PR DESCRIPTION
**Pull request by BradSwain**
I added a command line option to dump the preprocessed ir for debugging another issue. The flag `-dump-ir=mod.ll` will write the modified ir to `mod.ll`.

I also added a similar config for the json race report. The flag `-json=races.json` will write the json race report to `races.json`.

By default neither are written as most of the time we run the tool manually or in tests, we do not want/need the json or .ll files to be created.

### CODE REVIEW

File: src/RaceDetect.h [hunk 1]
```diff
@@ -15,6 +15,11 @@ limitations under the License.
 
 namespace race {
 
-Report detectRaces(llvm::Module *module);
+struct DetectRaceConfig {
```
Comment (funemy) [inline src/RaceDetect.h:18]: How about just call it `Config`?

File: src/main.cpp [hunk 2]
```diff
@@ -53,8 +63,11 @@ int main(int argc, char** argv) {
     llvm::outs() << race << "\n";
   }
   llvm::outs() << "Total Races Detected: " << report.size() << "\n";
-  report.dumpReport();
-  llvm::outs() << "JSON Report generated at ./races.json\n";
+
+  if (!DumpJSON.empty()) {
```
Comment (funemy) [inline src/main.cpp:67]: Why `DumpJson` is not a field of config?

Comment (BradSwain) [reply src/main.cpp:67]: Config is passed to `detectRaces` but writing json to a file is done in `main.cpp`, not `detectRaces`.

It does seem a little weird that IR dumping is done inside `detectRaces` but json dump is not. 

Maybe we should make both of them be in the same place.

File: src/RaceDetect.h (hunk 1 above)
Comment (BradSwain) [reply src/RaceDetect.h:18]: Prefer not because it is specific to detectRaces and is not under any specific namespace.

**Review by funemy — APPROVED** [review]
Not a super big deal at this time.

But I think we should keep all config options in a single data structure

**Pull request merged**